### DISCUSSION TITLE
Creating a new database per customer using Prisma Migration.

### DISCUSSION THREAD

**Discussion by HardikSheth1717** [prisma/prisma#18381] [Q&A] [answered]
Hello,

I am working as a Technical Lead in a product-based company. We are developing a multi-tenant (database per user) SaaS product. So, when a new user will sign up in our product, we are creating a new database for him/her using **Prisma migration**. This is working fine with 2-3 concurrent users. But if more than 10 users are trying to sign up then it will throw this connection limit error:

`Error: Migration engine error:
db error: FATAL: sorry, too many clients already`

Now, this will be a normal scenario when our product will live. We cannot restrict concurrent users from signing up :).

Right now, we are using below approach to create new tenant databases. We create a child process to run migration and create a new tenant database.

`execSync(
          export TENANT_DB_URL=${dbUrl} && npx prisma migrate deploy -- schema=${tenantSchemaPath}/tenant/schema.prisma, { stdio: "inherit" });`

I have also observed that our API is consuming too much RAM. Please check below screenshot:

![image](https://user-images.githubusercontent.com/12509977/226289255-ed5f8050-1999-4107-bd7b-f9a0c4d5d9c7.png)

As you can see API for login (userloveLoginMui with id = 0), is consuming 1.9 Gb of RAM which is too high.

So, I need few help / suggestions.

1. How to manage connections efficiently in multi-tenant applications?
2. What is the best approach to create tenant databases using **Prisma Migration**? I am thinking to create separate server or using AWS Lambda for migration purpose instead of using child process.
3. How to find route cause of high memory usage?

Tech stack we are using:
1. API - NodeJS with ExpressJS (REST API + GraphQL)
2. ORM - Prisma
3. Database - PostgreSQL
4. Front - end - ReactJS
5. Caching - Redis

Thanks :)

**Comment by nurul3101** [comment, marked as answer]
Hello @HardikSheth1717 👋 

> Error: Migration engine error: db error: FATAL: sorry, too many clients already

Can you share when exactly you get this error? How are you instantiating PrismaClient?


1. For efficiently managing connections, I would recommend you enable [metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) to get an idea of how many connections are being used by PrismaClient and what is the status for each of those connections. You can also change the connection pool size with [connection_limit](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#increasing-the-pool-size) parameter.

2. I would say that creating a separate server or service dedicated to handling migrations can help you isolate migration tasks from the main application. This way, the main application can continue serving requests while migrations are being processed.

3. You can enable [tracing](https://www.prisma.io/docs/concepts/components/prisma-client/opentelemetry-tracing) to get a detailed overview of every action performed by PrismaClient along with the time taken to perform that action. This could help you in identifying the bottlenecks and help in improving the performance.

**Reply by HardikSheth1717** [reply]
Hi @nurul3101 

I am getting this error when more than 10 users are trying to signup at the same time.

**Reply by HardikSheth1717** [reply]
Is it a good idea to implement queuing functionality for signup?

**Reply by nurul3101** [reply]
I think it would be a good idea, but can you share what exactly happens in the signup service?
Are you invoking this script in the signup service?
```ts
execSync( export TENANT_DB_URL=${dbUrl} && npx prisma migrate deploy -- schema=${tenantSchemaPath}/tenant/schema.prisma, { stdio: "inherit" });
```

**Reply by HardikSheth1717** [reply]
Yes, I am using above script to create tenant database by executing Prisma migration. Let me give you step by step details:

1. The user will fill up the signup form and hit Submit button.
2. The system will send an OTP on his/her email address.
3. As soon as he/she verifies the email address, the system will execute above script and create a new tenant database for him/her.

Now, I want to test performance of this flow. So, I asked everyone in my team to signup at the same time. We are a team of 20 people. When we did this thing, some of us were not able to signup. So, I checked the logs in PM2 and I found below error:

`Error: Migration engine error: db error: FATAL: sorry, too many clients already`

FYI, I have also implemented pgbouncer for connection pooling but Prisma migration does not work with pgbouncer.

I would also like to ask you about `execSync`. Is it correct way to execute migration using child process?

Thanks.

**Reply by nurul3101** [reply]
Thanks for providing the information.

Yes, for now, `execSync` can be used to spawn a process and execute migration. I would recommend you have a look at this GitHub Issue:
- https://github.com/prisma/prisma/issues/4688

There are some other alternatives mentioned in the issue, we do plan to support executing migrate commands through JavaScript/TypeScript.

**Reply by HardikSheth1717** [reply]
Thanks for the help!

I think for now I'll continue with the separate server and queuing for migrations. Let's see if it will work or not.

Thanks

**Reply by nurul3101** [reply]
No worries! 😄 

Do let us know in case it works or not 👍
